ethcore/engines/authority-round/src/randomness.rs

```diff
@@ -0,0 +1,245 @@
+//! On-chain randomness generation for authority round
+//!
+//! This module contains the support code for the on-chain randomness generation used by AuRa. Its
+//! core is the finite state machine `RandomnessPhase`, which can be loaded from the blockchain
+//! state, then asked to perform potentially necessary transaction afterwards using the `advance()`
+//! method.
+//!
+//! No additional state is kept inside the `RandomnessPhase`, it must be passed in each time.
+
+use derive_more::Display;
+use ethabi::Hash;
+use ethabi_contract::use_contract;
+use ethereum_types::{Address, U256};
+use keccak_hash::keccak;
+use log::{error, trace};
+use parity_crypto::publickey::{ecies, Error as CryptoError};
+use parity_bytes::Bytes;
+use rand::Rng;
+use engine::signer::EngineSigner;
+
+use crate::util::{BoundContract, CallError};
+
+/// Secret type expected by the contract.
+// Note: Conversion from `U256` back into `[u8; 32]` is cumbersome (missing implementations), for
+//       this reason we store the raw buffers.
+pub type Secret = [u8; 32];
```

Comment:
> I guess then we should change the method names in the contract as well, e.g. `reveal_secret`, etc.?

@afck, let's leave the names in the contracts as they are. I guess, David meant that `Secret` is already used in Parity's code, but I believe we should keep `RANDAO` terminology in the contracts.

> Using `H256` instead of a byte array is also something we'd need to change in the contract

Sorry, do you mean `bytes _cipher` in the `RandomAuRa.commitHash` function?

The [secret](https://github.com/poanetwork/posdao-contracts/blob/696fdd193319cb1afa21e7b96d3b82dbaec566c8/contracts/RandomAuRa.sol#L79) and [`currentSeed`](https://github.com/poanetwork/posdao-contracts/blob/696fdd193319cb1afa21e7b96d3b82dbaec566c8/contracts/RandomAuRa.sol#L28) in `RandomAuRa` contract are already `uint256`.

Reply:
Or you meant that the `secret` in the contract should be `bytes32`? If so, could you convert that to `uint256` on Parity's side to keep the contract compatible with our `aura-pos` fork?

Reply:
@varasev It is true that `Secret` is already a type in Parity, but I guess my objection was also around the role it plays in the RANDAO process (as I understood it, I might be completely wrong!): it is a secret piece of data for sure, but its main property is that it is the piece of randomness that this node contributes to the final random number. In other words: the name doesn't reflect its main purpose.

Reply:
@dvdplm Ok. Would you suggest renaming `revealSecret` to `revealNumber` in the contract?

Reply:
Right, sorry! I was confused because we're calling `revealSecret` with an `[u8; 32]` argument in the code, but that just works because it implements `Into<U256>`. The contract ABI is already fine then. :+1: 
(Apart from possibly the names.)

Reply:
Actually… `H256` would be better than `U256`. It's not interpreted as a number, so it has an [`as_fixed_bytes`](https://docs.rs/ethereum-types/0.7.0/ethereum_types/struct.H256.html#method.as_fixed_bytes) method and no endianness.

Reply:
@dvdplm I renamed `RandomAuRa.revealSecret` function to `revealNumber` in the code, in the contracts, and in my above comments:

https://github.com/paritytech/parity-ethereum/pull/10946/commits/eaa21315a5d43bba4661cb64ecc765a5c6ad7845
https://github.com/poanetwork/posdao-contracts/commit/ec3901160757ba295efd273a5cbd543037ceb1ce
https://github.com/paritytech/parity-ethereum/pull/10946#issuecomment-547857765
https://github.com/paritytech/parity-ethereum/pull/10946#issuecomment-548404697